File: java/replicant/src/main/java/replicant/ClientManager.java
package replicant;

import ch.qos.logback.classic.Logger;
import io.netty.channel.ChannelHandler.Sharable;
import io.netty.util.AttributeKey;
import multipaxos.MultiPaxos;
import org.slf4j.LoggerFactory;

import java.net.Socket;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

@Sharable
public class ClientManager {

    private static final Logger logger = (Logger) LoggerFactory.getLogger(ClientManager.class);
    private long nextId;
    private final long numPeers;
    private final MultiPaxos multiPaxos;
    private final AttributeKey<Long> clientIdAttrKey = AttributeKey.valueOf("ClientID");
    private final ReentrantLock mu = new ReentrantLock();
    private final HashMap<Long, Client> clients = new HashMap<>();

    public ClientManager(long id, long numPeers, MultiPaxos multiPaxos) {
        this.nextId = id;
        this.numPeers = numPeers;
        this.multiPaxos = multiPaxos;
    }

    private long nextClientId() {
        var id = nextId;
        nextId += numPeers;
        return id;
    }

    public void start(Socket socket) {
        var id = nextClientId();
        Client client = new Client(id, socket, multiPaxos, this);
        mu.lock();
        clients.put(id, client);
        mu.unlock();
        logger.debug("client_manager started client " + id);
        client.start();
    }

    public Client get(Long id) {
        mu.lock();
        try {
            return clients.get(id);
        } finally {
            mu.unlock();
        }
    }

    public void stop(Long id) {
        mu.lock();
        logger.debug("client_manager stopped client " + id);
        Client client = clients.get(id);
        assert (client != null);
        client.stop();
        clients.remove(id);
        mu.unlock();
    }

    public void stopAll() {
        mu.lock();
        for (Map.Entry<Long, Client> entry : clients.entrySet()) {
            Long id = entry.getKey();
            Client client = entry.getValue();
            logger.debug("client_manager stopping all clients " + id);
            client.stop();
            clients.remove(id);
        }
        mu.unlock();
    }
}

File: go-grpc/replicant/client_manager.go
package replicant

import (
	"github.com/psu-csl/replicated-store/go/multipaxos"
	logger "github.com/sirupsen/logrus"
	"net"
	"sync"
)

type ClientManager struct {
	nextId     int64
	numPeers   int64
	multipaxos *multipaxos.Multipaxos
	mu         sync.Mutex
	clients    map[int64]*Client
}

func NewClientManager(id int64,
	numPeers int64,
	mp *multipaxos.Multipaxos) *ClientManager {
	cm := &ClientManager{
		nextId:     id,
		numPeers:   numPeers,
		multipaxos: mp,
		clients:    make(map[int64]*Client),
	}
	return cm
}

func (cm *ClientManager) NextClientId() int64 {
	id := cm.nextId
	cm.nextId += cm.numPeers
	return id
}

func (cm *ClientManager) Start(socket net.Conn) {
	id := cm.NextClientId()
	client := NewClient(id, socket, cm.multipaxos, cm)

	cm.mu.Lock()
	cm.clients[id] = client
	cm.mu.Unlock()
	logger.Infof("client_manager started client %v\n", id)
	client.Start()
}

func (cm *ClientManager) Get(id int64) *Client {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	client, ok := cm.clients[id]
	if !ok {
		return nil
	}
	return client
}

func (cm *ClientManager) Stop(id int64) {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	logger.Infof("client_manager stopped client %v\n", id)
	client, ok := cm.clients[id]
	if !ok {
		panic("no client to stop")
	}
	client.Stop()
	delete(cm.clients, id)
}

func (cm *ClientManager) StopAll() {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	for id, client := range cm.clients {
		logger.Infof("client_manager stopping all clients %v\n", id)
		client.Stop()
		delete(cm.clients, id)
	}
}

func (cm *ClientManager) NumClients() int {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	return len(cm.clients)
}
